pkg/daemon/update.go

```diff
@@ -1561,20 +1419,17 @@ func newMachineConfigDiff(oldConfig, newConfig *mcfgv1.MachineConfig) (*machineC
 		units:      !reflect.DeepEqual(oldIgn.Systemd.Units, newIgn.Systemd.Units),
 		kernelType: canonicalizeKernelType(oldConfig.Spec.KernelType) != canonicalizeKernelType(newConfig.Spec.KernelType),
 		extensions: !(extensionsEmpty || reflect.DeepEqual(oldConfig.Spec.Extensions, newConfig.Spec.Extensions)),
-	}, nil
-}
+		oclEnabled: (oldOCLImage != "" || newOCLImage != "") || (oldOCLImage != "" && newOCLImage != ""),
+	}
 
-func newMachineConfigDiffFromLayered(oldConfig, newConfig *mcfgv1.MachineConfig, oldImage, newImage string) (*machineConfigDiff, error) {
-	mcDiff, err := newMachineConfigDiff(oldConfig, newConfig)
-	if err != nil {
-		return mcDiff, err
+	if !diff.oclEnabled {
+		return diff, nil
 	}
 
-	mcDiff.oclEnabled = true
-	// If the new OS image is empty, that means we are in a revert-from-OCL situation.
-	mcDiff.revertFromOCL = newImage == ""
-	mcDiff.osUpdate = oldImage != newImage || forceFileExists()
-	return mcDiff, nil
+	// If OCL is enabled, compute OS update and revertFromOCL separately.
+	diff.osUpdate = oldOCLImage != newOCLImage || force
+	diff.revertFromOCL = newOCLImage == ""
```

Comment:
thought (non-blocking): We can tighten up this check by doing something like `diff.revertFromOCL = newOCLImage == "" && oldOCLImage != ""`.